blueflood-rollupTools/src/main/java/com/rackspacecloud/blueflood/CloudFilesBackfiller/rollup/handlers/BuildStore.java

```diff
@@ -100,24 +96,52 @@ public static void merge (InputStream jsonInput) throws IOException {
                     continue;
                 }
 
-                // todo: per-line instrumentation starts here.
-
                 CheckFromJson checkFromJson = gson.fromJson(line, CheckFromJson.class);
+
+                long timestamp = checkFromJson.getTimestamp();
+                long snappedMillis = Granularity.MIN_5.snapMillis(timestamp);
+
+                Range rangeOfThisTimestamp = new Range(snappedMillis, snappedMillis + Granularity.MIN_5.milliseconds() - 1);
+
+                // Do not add timestamps that lie out of range
+                if (!rangesToRollup.contains(rangeOfThisTimestamp)) {
+                    log.warn("Timestamp of metrics of check found lying out the range: "+rangeOfThisTimestamp+" TS: "+timestamp);
+                    line = reader.readLine();
+                    continue;
+                }
+
+                //These are out of band timestamps lying in the ranges which we have already rolled
+                if (rangesToRollup.contains(rangeOfThisTimestamp) && !rangesStillApplicable.contains(rangeOfThisTimestamp)) {
+                    log.warn("Range of timestamp of metrics of check " + checkFromJson.getCheckId() + "is out of applicable ranges");
+                    outOfRangeToleration++;
+
+                    // If we are seeing a lot of out of band metrics, something is wrong. May be metrics are back logged a lot. stop immediately. try to increase the range buffer?
+                    if (outOfRangeToleration > OUT_OF_RANGE_TOLERATION_THRESHOLD) {
+                        throw new OutOFBandException("Starting to see a lot of metrics in non-applicable ranges");
+                    }
+                    line = reader.readLine();
+                    continue;
+                }
+
```

Comment:
instead of moving the timestamp check out of the metrics loop, would it be better to check each individual metrics timestamps?